Sources/AsyncAlgorithms/AsyncSharedSequence.swift

```diff
@@ -0,0 +1,578 @@
+//
+// This source file is part of the Swift Async Algorithms open source project
+//
+// Copyright (c) 2022 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+import DequeModule
+
+extension AsyncSequence {
+  
+  /// Creates an asynchronous sequence that can be shared by multiple consumers.
+  ///
+  /// - parameter history: the number of previously emitted elements to prefix to the iterator of a new
+  ///   consumer
+  /// - parameter iteratorDisposalPolicy:the iterator disposal policy applied by a shared
+  ///   asynchronous sequence to its upstream iterator
+  public func share(
+    history historyCount: Int = 0,
+    disposingBaseIterator iteratorDisposalPolicy: AsyncSharedSequence<Self>.IteratorDisposalPolicy = .whenTerminatedOrVacant
+  ) -> AsyncSharedSequence<Self> {
+    AsyncSharedSequence(
+      self, history: historyCount, disposingBaseIterator: iteratorDisposalPolicy)
+  }
+}
+
+// MARK: - Sequence
+
+/// An asynchronous sequence that can be iterated by multiple concurrent consumers.
+///
+/// Use a shared asynchronous sequence when you have multiple downstream asynchronous sequences
+/// with which you wish to share the output of a single asynchronous sequence. This can be useful if
+/// you have expensive upstream operations, or if your asynchronous sequence represents the output
+/// of a physical device.
+///
+/// Elements are emitted from a shared asynchronous sequence at a rate that does not exceed the
+/// consumption of its slowest consumer. If this kind of back-pressure isn't desirable for your
+/// use-case, ``AsyncSharedSequence`` can be composed with buffers – either upstream, downstream,
+/// or both – to acheive the desired behavior.
+///
+/// If you have an asynchronous sequence that consumes expensive system resources, it is possible to
+/// configure ``AsyncSharedSequence`` to discard its upstream iterator when the connected
+/// downstream consumer count falls to zero. This allows any cancellation tasks configured on the
+/// upstream asynchronous sequence to be initiated and for expensive resources to be terminated.
+/// ``AsyncSharedSequence`` will re-create a fresh iterator if there is further demand.
+///
+/// For use-cases where it is important for consumers to have a record of elements emitted prior to
+/// their connection, a ``AsyncSharedSequence`` can also be configured to prefix its output with
+/// the most recently emitted elements. If ``AsyncSharedSequence`` is configured to drop its
+/// iterator when the connected consumer count falls to zero, its history will be discarded at the
+/// same time.
+public struct AsyncSharedSequence<Base: AsyncSequence>
+where Base: Sendable, Base.Element: Sendable, Base.AsyncIterator: Sendable {
```

Comment:
I still need to read through all of this but we cannot work with the requirement that the `Base.AsyncIterator` is `Sendable`. `AsyncIterator`s should almost never be `Sendable` so this algo wouldn't be usable

Reply:
Got it. Well, I *think* it should be fine if we use `@unchecked` on a particular portion of the state as the iterator is being used in a particular way that prevents concurrent access. I'll need someone to confirm my understanding on that though!

I've remove the constraint, added `@unchecked` where appropriate and run through the compiler via `swift build -Xswiftc -Xfrontend -Xswiftc -strict-concurrency=complete` and I'm no longer seeing any warnings, so fingers-crossed this way of using an iterator is viable.